### PR TITLE
Backport modern quadratic views to deprecated BQMs

### PR DESCRIPTION
**Pull request by arcondello**

### CODE REVIEW

File: dimod/core/bqm.py [hunk 1]
```diff
@@ -897,6 +668,27 @@ def min_and_max(iterable):
         else:
             return 1.0
 
+    def reduce_linear(self, function, initializer=None):
+        gen = (self.get_linear(v) for v in self.variables)
+        if initializer is None:
+            return functools.reduce(function, gen)
+        else:
+            return functools.reduce(function, gen, initializer)
```
Comment (randomir) [inline dimod/core/bqm.py:676]:
```suggestion
    def reduce_linear(self, function, *args):
        gen = (self.get_linear(v) for v in self.variables)
        return functools.reduce(function, gen, *args)
```
Shorter, with initializer undocumented, but it also accepts `None` for initial value.

Comment (randomir) [reply dimod/core/bqm.py:676]: Or
```suggestion
    def reduce_linear(self, function, initializer=_value_missing):
        gen = (self.get_linear(v) for v in self.variables)
        if initializer is _value_missing:
            return functools.reduce(function, gen)
        else:
            return functools.reduce(function, gen, initializer)
```
with
```
_value_missing = object()
```
in module scope.

**Review by randomir — APPROVED** [review]
LGTM.

File: dimod/core/bqm.py (hunk 1 above)
Comment (arcondello) [reply dimod/core/bqm.py:676]: In this case the biases are known to be numbers, which is why I was using `None`. But suppose `DictBQM` supports arbitrary biases so may as well use the `_value_missing` version.

Comment (arcondello) [reply dimod/core/bqm.py:676]: Once I started pulling on this thread I found some other stuff I wanted to do, so  I am going to make this change in a followup PR

**Pull request merged**